becquerel/core/autocal.py

```diff
@@ -0,0 +1,267 @@
+                comb_chan = comb_chan[ind]
+                comb_snr = comb_snr[ind]
+                # calculate gain
+                gain = fit_gain(comb_chan, comb_snr, comb_erg)
+                if gain < gain_range[0] or gain > gain_range[1]:
+                    continue
+                # calculate predicted energies
+                pred_erg = gain * comb_chan
+                de = pred_erg - comb_erg
+                if (abs(de) > de_max).any():
+                    continue
+                # calculate figure of merit
+                fom = fom_gain(comb_chan, comb_snr, comb_erg)
+                if verbose:
+                    print('v')
+                    s0 = 'Valid calibration found:\n'
+                    s0 += 'FOM: {:15.9f}'.format(fom)
+                    s0 += '  gain: {:6.3f}'.format(gain)
+                    s0 += '  ergs: {:50s}'.format(str(comb_erg))
+                    s0 += '  de: {:50s}'.format(str(de))
+                    s0 += '  chans: {:40s}'.format(str(comb_chan))
+                    print(s0)
+                if best_fom is None:
+                    best_fom = fom + 1.
+                if fom < best_fom:
+                    best_fom = fom
+                    best_gain = gain
+                    best_chans = comb_chan
+                    best_snrs = comb_snr
+                    best_ergs = comb_erg
+                    if verbose:
+                        s0 = 'Best calibration so far:\n'
+                        s0 += 'FOM: {:15.9f}'.format(best_fom)
+                        s0 += '  gain: {:6.3f}'.format(best_gain)
+                        s0 += '  ergs: {:50s}'.format(str(best_ergs))
+                        s0 += '  de: {:50s}'.format(str(de))
+                        s0 += '  chans: {:40s}'.format(str(best_chans))
+                        print(s0)
+        n_set -= 1
+    if best_gain is None:
+        return None
+    else:
+        return {
+            'gain': best_gain,
+            'channels': best_chans,
+            'snrs': best_snrs,
+            'energies': best_ergs,
+        }
+
+
+class AutoCalibratorError(Exception):
+    """Base class for errors in AutoCalibrator."""
+
+    pass
+
+
+class AutoCalibrator(object):
+    """Automatically calibrate a spectrum by convolving it with a filter."""
+
+    def __init__(self, peakfinder):
+        """Initialize the calibration with a spectrum and kernel."""
+        self.set_peaks(peakfinder)
+        self.gain = None
+        self.cal = None
+        self.fit_channels = []
+        self.fit_snrs = []
+        self.fit_energies = []
+        self.reset()
+
+    def reset(self):
+        """Reset all of the members."""
+        self.gain = None
+        self.cal = None
+        self.fit_channels = []
+        self.fit_snrs = []
+        self.fit_energies = []
+
+    def set_peaks(self, peakfinder):
+        """Use the peaks found by the PeakFinder."""
+        assert isinstance(peakfinder, PeakFinder)
```

Comment:
@markbandstra can you change these from `assert` to raise an `AutoCalibrationError`?